README
==========
SPADE
==========


Welcome to SPADE
===========================
SPADE (Smart Python multi-Agent Development Environment) is a Multiagent and Organizations Platform based on the XMPP/Jabber technology and written in the Python programming language. This technology offers by itself many features and facilities that ease the construction of MAS, such as an existing communication channel, the concepts of users (agents) and servers (platforms) and an extensible communication protocol based on XML, just like FIPA-ACL. Many other agent platforms exist, but SPADE is the first to base its roots on the XMPP technology.

The SPADE Agent Platform does not require (but strongly recommends) the operation of agents made with the SPADE Agent Library. The platform itself uses the library to empower its internals, but aside from that, you can develop your own agents in the programming language of your choice and use them with SPADE. The only requirement those agents must fulfill is to be able to communicate through the XMPP protocol. The FIPA-ACL messages will be embedded in XMPP messages. Be warned, however, that some features of the whole SPADE experience may not be available if you do not use the SPADE Agent Library to build your agents.

The SPADE Agent Library
============================
The SPADE Agent Library is a module for the Python programming language for building SPADE agents. It is a collection of classes, functions and tools for creating new SPADE agents that can work with the SPADE Agent Platform. Using it is the best way to start developing your own SPADE agents. In the future, we would like to offer bindings of the SPADE Agent Library for more programming languages, like Java or C#, but for now only Python is supported.

If you have worked with Python before, you will find the SPADE Agent Library easy to understand and use, with classes ready to use and expand, and many functionalities already built in the library. If you have worked with other high-level behavioral agent platforms (like Jade or MadKit) you will find some similarities in the way the library works, although the agent model present in SPADE is a bit different.


ChangeLog
2012-11-23 14:16  jpalanca

	* trunk/*: (1) xmppd/xmppd.py: Solved multisession_thread daemonic bug.
	  (2) spade/Agent.py: Solved kill issues whith daemonic threads,
	  added hasBehaviour method, if register fails AgentNotRegisteredError 
	  exception is raised. 
	  (3) spade/MessageReceiver.py: _receive is not blocked anymore, 
	  checking periodically if was killed. 
	  (4) test/kbTestCase.py: skipped SPARQL tests. 
	  (5) spade/mtp/http.py: solved blocking thread. Now has a timeout. 
	  (6) spade/peer2peer.py: P2P behavior is also no longer blocking. Uses a timeout. 
	  (7) spade/wui.py: another daemonic thread bug solved. 
	  (8) configure.py: Added default admin password for WUI. The default password is 'secret'. 
	  (9) All files have been modified according to the PEP8 conventions. 
	  (10) All files encoded in UTF-8. 
	  This is the last commit using subversion. We switch to git. Pray what you know...

2012-11-12 17:28  jpalanca

	* trunk/doc/manual/book.xml,
	  trunk/doc/manual/ch00-introduction-to-spade.xml,
	  trunk/doc/manual/ch01-quickstart-guide.xml,
	  trunk/doc/manual/ch02-fundamental-concepts.xml,
	  trunk/doc/manual/ch03-basic-agents.xml,
	  trunk/doc/manual/ch06-bdi.xml, trunk/doc/manual/html/ch05s03.html,
	  trunk/doc/manual/html/index.html,
	  trunk/doc/manual/html/spade.advanced.eventbehaviour.html,
	  trunk/doc/manual/html/spade.advanced.fsm.html,
	  trunk/doc/manual/html/spade.advanced.html,
	  trunk/doc/manual/html/spade.basic.html,
	  trunk/doc/manual/html/spade.basic.im.html,
	  trunk/doc/manual/html/spade.basic.mas.html,
	  trunk/doc/manual/html/spade.basic.spadelibrary.html,
	  trunk/doc/manual/html/spade.basic.spadeplatform.html,
	  trunk/doc/manual/html/spade.basic.whatsnext.html,
	  trunk/doc/manual/html/spade.basicagents.agentmodel.communication.html,
	  trunk/doc/manual/html/spade.basicagents.agentmodel.html,
	  trunk/doc/manual/html/spade.basicagents.behav.html,
	  trunk/doc/manual/html/spade.basicagents.first.html,
	  trunk/doc/manual/html/spade.basicagents.html,
	  trunk/doc/manual/html/spade.basicagents.library.html,
	  trunk/doc/manual/html/spade.basicagents.whatsnext.html,
	  trunk/doc/manual/html/spade.bdi.future.html,
	  trunk/doc/manual/html/spade.bdi.goals.html,
	  trunk/doc/manual/html/spade.bdi.html,
	  trunk/doc/manual/html/spade.bdi.intro.html,
	  trunk/doc/manual/html/spade.bdi.kb.html,
	  trunk/doc/manual/html/spade.bdi.plans.html,
	  trunk/doc/manual/html/spade.bdi.running.html,
	  trunk/doc/manual/html/spade.foreword.html,
	  trunk/doc/manual/html/spade.fundamental.html,
	  trunk/doc/manual/html/spade.fundamental.platform.html,
	  trunk/doc/manual/html/spade.platform.ams.html,
	  trunk/doc/manual/html/spade.platform.df.html,
	  trunk/doc/manual/html/spade.platform.html,
	  trunk/doc/manual/html/spade.quickstart.html,
	  trunk/doc/manual/html/spade.quickstart.installation.html,
	  trunk/doc/manual/html/spade.quickstart.requirements.html,
	  trunk/setup.py, trunk/test/bdiTestCase.py,
	  trunk/test/kbTestCase.py: - Updated documentation with corrections
	  about the SPADE 2.1 release. - Rewritten the BDI chapter of the
	  documentation, now with the current implementation of the bdi
	  module. - Updated SPADE requirements. Now SPADE needs Python 2.7
	  or above (sorry old machines).

2012-11-09 18:06  jpalanca

	* trunk/test/httpTestCase.py, trunk/test/jsonTestCase.py: added
	  missing testcases

2012-11-09 17:51  jpalanca

	* trunk/configure.py, trunk/runspade.py, trunk/spade/ACLMessage.py,
	  trunk/spade/ACLParser.py, trunk/spade/AID.py,
	  trunk/spade/Agent.py, trunk/spade/Envelope.py, trunk/spade/MTP.py,
	  trunk/spade/Platform.py, trunk/spade/mtp/http.py,
	  trunk/spade/spade_backend.py, trunk/spade/wui.py, trunk/test.py,
	  trunk/test/kbTestCase.py, trunk/test/unittests_spade.xml: Version
	  2.1: -JSON support for ACL Messages and Envelopes -HTTP protocol
	  can send both XML and JSON messages -New unittests added:
	  jsonTestCase and httpTestCase -Some minor bugs fixed -HTTP
	  transport is now activated by default -XMPPD server deals with
	  max_sockets error -AID, ACLMessage and Envelope classess can be
	  marshalled in JSON (asJSON) and unmarshalled from JSON (loadJSON)

2012-10-22 21:29  schatten.markus@gmail.com

	* trunk/xmpp/auth.py: Fixed spelling error in xmpp/auth.py

2012-08-01 16:48  jpalanca

	* trunk/doc/api, trunk/doc/api/api-objects.txt,
	  trunk/doc/api/class-tree.html, trunk/doc/api/crarr.png,
	  trunk/doc/api/epydoc.css, trunk/doc/api/epydoc.js,
	  trunk/doc/api/frames.html, trunk/doc/api/help.html,
	  trunk/doc/api/identifier-index-A.html,
	  trunk/doc/api/identifier-index-B.html,
	  trunk/doc/api/identifier-index-C.html,
	  trunk/doc/api/identifier-index-D.html,
	  trunk/doc/api/identifier-index-E.html,
	  trunk/doc/api/identifier-index-F.html,
	  trunk/doc/api/identifier-index-G.html,
	  trunk/doc/api/identifier-index-H.html,
	  trunk/doc/api/identifier-index-I.html,
	  trunk/doc/api/identifier-index-J.html,
	  trunk/doc/api/identifier-index-K.html,
	  trunk/doc/api/identifier-index-L.html,
	  trunk/doc/api/identifier-index-M.html,
	  trunk/doc/api/identifier-index-N.html,
	  trunk/doc/api/identifier-index-O.html,
	  trunk/doc/api/identifier-index-P.html,
	  trunk/doc/api/identifier-index-Q.html,
	  trunk/doc/api/identifier-index-R.html,
	  trunk/doc/api/identifier-index-S.html,
	  trunk/doc/api/identifier-index-T.html,
	  trunk/doc/api/identifier-index-U.html,
	  trunk/doc/api/identifier-index-V.html,
	  trunk/doc/api/identifier-index-W.html,
	  trunk/doc/api/identifier-index-X.html,
	  trunk/doc/api/identifier-index-Y.html,
	  trunk/doc/api/identifier-index-Z.html,
	  trunk/doc/api/identifier-index-_.html,
	  trunk/doc/api/identifier-index.html, trunk/doc/api/index.html,
	  trunk/doc/api/mimetools.Message-class.html,
	  trunk/doc/api/module-tree.html, trunk/doc/api/redirect.html,
	  trunk/doc/api/spade-module.html, trunk/doc/api/spade-pysrc.html,
	  trunk/doc/api/spade.ACLMessage-module.html,

recv.py
#!/usr/bin/env python

# $Id: recv.py,v 1.1 2004/02/06 20:48:49 jpalanca Exp $

# You may need to change the above line to point at
# python rather than python2 depending on your os/distro

import socket
from select import select
from string import split,strip,join,find
import sys,os
import smtplib
import telnetlib
import pprint
pp = pprint.PrettyPrinter(indent=4)

sys.path.insert(1, os.path.join(sys.path[0], '..'))

import jabber

True = 1
False = 0

# Change this to 0 if you dont want a color xterm
USE_COLOR = 1

Who = ''
MyStatus = ''
MyShow   = ''

def XMLdecode(body):
	
    tag = getTag(body)
    if tag == "REQUEST":
        print "He recibido un REQUEST"
        
    return
def tellIQ(fromwho, agent, con, me):
	if fromwho != '':
            i = jabber.Iq(agent)
	    i.setType('set')
	    i.setFrom(me)
	    i.setID("search2")
	    i.setQuery("jabber:iq:search")
	    i.setQueryPayload("<last>Agent</last>")
	    pp.pprint(i)
            con.send(i)
	    '''
	    i.setType('set')
	    i.setFrom(me)
	    i.setID('register2')
	    i.setQuery("jabber:iq:register")
	    i.setQueryPayload("<query><first>PROVEMAN</first><last>FipperAgent</last><nick>AGENTONE</nick><email>a1@fipper.org</email></query>")
	    pp.pprint(i)
	    con.send(i)
	    '''
	return
	
def tellFortune(fromwho):
	os.system("fortune > out.txt")
	fsock = open('out.txt', 'r')
	s = fsock.read()
	fromwho = "/select "+str(fromwho)
	doCmd(con,fromwho)
	doCmd(con, s)
	fsock.close()
	os.system("rm out.txt")
	return

def tellGoogle(fromwho, what):
	that = 'links -dump "http://www.google.es/search?q='+what+'" > out-gg.txt'
	print that
	os.system(that)
	fsock = open('out-gg.txt', 'r')
	s = fsock.read()
	fromwho = "/select "+str(fromwho)
	doCmd(con,fromwho)
	doCmd(con, s)
	fsock.close()
	os.system("rm out-gg.txt")
	return
	
def tellWikipedia(fromwho, what):
	that = 'links -dump "http://en.wikipedia.org/w/wiki.phtml?search='+what+'&go=Go" > out-wp.txt'
	print that
	os.system(that)
	fsock = open('out-wp.txt', 'r')
	s = fsock.read()
	fromwho = "/select "+str(fromwho)
	doCmd(con,fromwho)
	doCmd(con, s)
	fsock.close()
	os.system("rm out-wp.txt")
	return
	
def tellMLD(fromwho):
	tn = telnetlib.Telnet("192.168.1.2", "4000")
	#~ tn.read_until(">")
	tn.write("ansi false\n")
	tn.write("vd\n")
	tn.write("q\n")
	thing = tn.read_until("blobloblo",2) # Dirty hack from hell
	i = thing.find("Downloaded")
	j = thing.rfind("Downloaded")
	thing = thing[i:j]
	fromwho = "/select "+str(fromwho)
	doCmd(con,fromwho)
	doCmd(con,str(thing))
	return
	
def tellMail(fromwho, maildata):
	md = split(maildata, sep=';;')
	print md
	if len(md) != 3:
		fromwho = "/select "+str(fromwho)
		doCmd(con,fromwho)
		doCmd(con,"MAIL SYNTAX ERROR")
	else:
		fw = fromwho
		server = smtplib.SMTP('maulo.ma.cx')
		msg = "Subject: "+str(md[1])+"\n"+str(md[2])
		i = find(str(fromwho),"/")
		fromwho = str(fromwho)[0:i]
		server.sendmail(str(fromwho), md[0], msg)
		server.quit()
		fw = "/select "+str(fw)
		doCmd(con,fw)
		doCmd(con,"Gracias por usar nuestro servicio de correo.\nPor favor, recuerde no abusar de ello.")
	return

def usage():
    print "%s: a simple python jabber client " % sys.argv[0]
    print "usage:"
    print "%s <server> - connect to <server> and register" % sys.argv[0]
    print "%s server> <username> <password> <resource>"    % sys.argv[0]
    print "            - connect to server and login   "
    sys.exit(0)


def doCmd(con,txt):
    global Who
    if txt[0] == '/' :
        cmd = split(txt)
        if cmd[0] == '/select':
            Who = cmd[1]
            print "%s selected" % cmd[1]
        elif cmd[0] == '/presence':
            to = cmd[1]
            type = cmd[2]
            con.send(jabber.Presence(to, type))
        elif cmd[0] == '/status':
            p = jabber.Presence()
            MyStatus = ' '.join(cmd[1:])
            p.setStatus(MyStatus)
            con.send(p)
        elif cmd[0] == '/show':
            p = jabber.Presence()
            MyShow = ' '.join(cmd[1:])
            p.setShow(MyShow)
            con.send(p)
        elif cmd[0] == '/subscribe':
            to = cmd[1]
            con.send(jabber.Presence(to, 'subscribe'))
        elif cmd[0] == '/unsubscribe':
            to = cmd[1]
            con.send(jabber.Presence(to, 'unsubscribe'))
        elif cmd[0] == '/roster':
            con.requestRoster()
            _roster = con.getRoster()
            for jid in _roster.getJIDs():
                print colorize("%s :: %s (%s/%s)" 
                               % ( jid, _roster.getOnline(jid),
                                   _roster.getStatus(jid),
                                   _roster.getShow(jid),
                                   ) , 'blue' )

        elif cmd[0] == '/agents':
            print con.requestAgents()
        elif cmd[0] == '/register':
            agent = ''
            if len(cmd) > 1:
                agent = cmd[1]
            con.requestRegInfo(agent)
            print con.getRegInfo()
	elif cmd[0] == "/iq":
	    tellIQ(JID, cmd[1], con, JID)
        elif cmd[0] == '/exit':
            con.disconnect()
            print colorize("Bye!",'red')
            sys.exit(0)
        elif cmd[0] == '/help':
            print "commands are:"
            print "   /select <jabberid>"
            print "      - selects who to send messages to"
            print "   /subscribe <jid>"
            print "      - subscribe to jid's presence"
            print "   /unsubscribe <jid>"
            print "      - unsubscribe to jid's presence"
            print "   /presence <jabberid> <type>"
            print "      - sends a presence of <type> type to the jabber id"
            print "   /status <status>"
            print "      - set your presence status message"
            print "   /show <status>"
            print "      - set your presence show message"
            print "   /roster"
            print "      - requests roster from the server and "
            print "        display a basic dump of it."
            print "   /exit"
            print "      - exit cleanly"
        else:
            print colorize("uh?", 'red')
    else:
        if Who != '':
            msg = jabber.Message(Who, strip(txt))
            msg.setType('chat')
            print "<%s> %s" % (JID, msg.getBody())
            con.send(msg)
        else:
            print colorize('Nobody selected','red')
            

def messageCB(con, msg):
    """Called when a message is recieved"""
    if msg.getBody(): ## Dont show blank messages ##
        print colorize(
            '<' + str(msg.getFrom()) + '>', 'green'
            ) + ' ' + msg.getBody()
    
    #~ XMLdecode(msg.getBody())
    
    cuerpo = msg.getBody()
    if cuerpo == "CAPTAIN":
        doCmd(con,"WHAT_HAPPEN")
    elif cuerpo == "WHAT_HAPPEN":
	doCmd(con,"WE_GET_SIGNAL")
    elif cuerpo == "WE_GET_SIGNAL":
	doCmd(con,"MAIN_SCREEN_TURN_ON")
    elif cuerpo == "MAIN_SCREEN_TURN_ON":
	doCmd(con,"ITS_YOU")
    elif cuerpo == "ITS_YOU":
	doCmd(con,"HOW_ARE_YOU_GENTLEMEN")
    elif cuerpo == "HOW_ARE_YOU_GENTLEMEN":
	doCmd(con,"ALL_YOUR_BASE_ARE_BELONG_TO_US")
    elif cuerpo == "ALL_YOUR_BASE_ARE_BELONG_TO_US":
	doCmd(con,"WHAT_YOU_SAY")
    elif cuerpo == "FORTUNE":
	tellFortune(msg.getFrom())
    elif cuerpo[0:6] == "GOOGLE":
	    cuerpo = cuerpo[7:]
	    tellGoogle(msg.getFrom(),cuerpo)
    elif cuerpo[0:2] == "WP":
	    cuerpo = cuerpo[3:]
	    tellWikipedia(msg.getFrom(), cuerpo)
    elif cuerpo[0:4] == "MAIL":
	    cuerpo = cuerpo[5:]
	    tellMail(msg.getFrom(), cuerpo)
    elif cuerpo == "MLD":
	    tellMLD(msg.getFrom())

    return

def presenceCB(con, prs):
    """Called when a presence is recieved"""
    who = str(prs.getFrom())
    type = prs.getType()
    if type == None: type = 'available'

    # subscription request: 
    # - accept their subscription
    # - send request for subscription to their presence
    if type == 'subscribe':
        print colorize("subscribe request from %s" % (who), 'blue')
        con.send(jabber.Presence(to=who, type='subscribed'))
        con.send(jabber.Presence(to=who, type='subscribe'))

    # unsubscription request: 
    # - accept their unsubscription
    # - send request for unsubscription to their presence
    elif type == 'unsubscribe':
        print colorize("unsubscribe request from %s" % (who), 'blue')
        con.send(jabber.Presence(to=who, type='unsubscribed'))
        con.send(jabber.Presence(to=who, type='unsubscribe'))

    elif type == 'subscribed':
        print colorize("we are now subscribed to %s" % (who), 'blue')

    elif type == 'unsubscribed':
        print colorize("we are now unsubscribed to %s"  % (who), 'blue')

    elif type == 'available':
        print colorize("%s is available (%s / %s)" % \
                       (who, prs.getShow(), prs.getStatus()),'blue')
    elif type == 'unavailable':
        print colorize("%s is unavailable (%s / %s)" % \
                       (who, prs.getShow(), prs.getStatus()),'blue')


def iqCB(con,iq):
    """Called when an iq is recieved, we just let the library handle it at the moment"""
    print iq

def disconnectedCB(con):
    print colorize("Ouch, network error", 'red')
    sys.exit(1)

def colorize(txt, col):
    """Return colorized text"""
    if not USE_COLOR: return txt ## DJ - just incase it breaks your terms ;) ##
    cols = { 'red':1, 'green':2, 'yellow':3, 'blue':4}
    initcode = '\033[;3'
    endcode  = '\033[0m'
    if type(col) == type(1): 
        return initcode + str(col) + 'm' + txt + endcode
    try: return initcode + str(cols[col]) + 'm' + txt + endcode
    except: return txt


if len(sys.argv) == 1: usage()
Server = sys.argv[1]
Username = ''
Password = ''
Resource = 'default'


con = jabber.Client(host=Server,debug=False ,log=sys.stderr)

# Experimental SSL support
#
# con = jabber.Client(host=Server,debug=True ,log=sys.stderr,
#                    port=5223, connection=xmlstream.TCP_SSL)

try:
    con.connect()
except IOError, e:
    print "Couldn't connect: %s" % e
    sys.exit(0)
else:
    print colorize("Connected",'red')

con.setMessageHandler(messageCB)
con.setPresenceHandler(presenceCB)
con.setIqHandler(iqCB)
con.setDisconnectHandler(disconnectedCB)

if len(sys.argv) == 2:
    # Set up a jabber account
    con.requestRegInfo()
    req = con.getRegInfo()
    print req[u'instructions']
    for info in req.keys():
        if info != u'instructions' and \
           info != u'key':
            print "enter %s;" % info
            con.setRegInfo( info,strip(sys.stdin.readline()) )
    con.sendRegInfo()
    req = con.getRegInfo()
    Username = req['username']
    Password = req['password']
else:
    Username = sys.argv[2]
    Password = sys.argv[3]
    Resource = sys.argv[4]

print colorize("Attempting to log in...", 'red')


if con.auth(Username,Password,Resource):
    print colorize("Logged in as %s to server %s" % ( Username, Server), 'red')
else:
    print "eek -> ", con.lastErr, con.lastErrCode
    sys.exit(1)

print colorize("Requesting Roster Info" , 'red')
con.requestRoster()
con.sendInitPresence()
print colorize("Ok, ready" , 'red')
print colorize("Type /help for help", 'red')

JID = Username + '@' + Server + '/' + Resource

while(1):
    inputs, outputs, errors = select([sys.stdin], [], [],1)

    if sys.stdin in inputs:
        doCmd(con,sys.stdin.readline())
    else:
        con.process(1)
    
        









